backend/TraderX/src/main/java/cmpe451/group6/rest/transaction/service/TransactionService.java

```diff
@@ -0,0 +1,204 @@
+
+    public boolean buyAsset(String requesterName, String code, float amount) {
+        User user = userRepository.findByUsername(requesterName);
+        if (user == null) {
+            throw new CustomException("There is no user named " + requesterName + ".", HttpStatus.NOT_ACCEPTABLE);
+        }
+        Equipment equipment = equipmentRepository.findByCode(code);
+        if (equipment == null) {
+            throw new CustomException("There is no equipment with code " + code + ".", HttpStatus.NOT_ACCEPTABLE);
+        }
+
+        Asset asset = assetRepository.getAsset(requesterName, EquipmentConfig.BASE_CURRENCY_CODE);
+        if (asset == null) {
+            throw new CustomException("The user has no money in application.", HttpStatus.NOT_ACCEPTABLE);
+        }
+
+        float neededMoney = (float) (amount * equipment.getCurrentValue());
+        float usersMoney = asset.getAmount();
+        if (neededMoney > usersMoney) {
+            throw new CustomException("The user doesn't have enough money to buy " + amount + " " + code + ".",
+                    HttpStatus.NOT_ACCEPTABLE);
+        }
+
+        // CAN BUY
+
+        // set remaining money
+        asset.setAmount(usersMoney - neededMoney);
+        assetRepository.save(asset);
+
+        // check if the user already had asset of that type
+        Asset alreadyHad = assetRepository.getAsset(requesterName, equipment.getCode());
+        if (alreadyHad == null) {
+            // set newly bought asset
+            Asset temp = new Asset();
+            temp.setAmount(amount);
+            temp.setUser(user);
+            temp.setEquipment(equipmentRepository.findByCode(code));
+            assetRepository.save(temp);
+        } else {
+            alreadyHad.setAmount(alreadyHad.getAmount() + amount);
+            assetRepository.save(alreadyHad);
+        }
+
+        Transaction newTransaction = new Transaction();
+        newTransaction.setUser(user);
+        newTransaction.setEquipment(equipment);
+        newTransaction.setTransactionType(TransactionType.BUY);
+        TransactionRepository.save(newTransaction);
+        return true;
+    }
+
+    public boolean sellAsset(String requesterName, String code, float amount) {
+        User user = userRepository.findByUsername(requesterName);
+        if (user == null) {
+            throw new CustomException("There is no user named " + requesterName + ".", HttpStatus.NOT_ACCEPTABLE);
+        }
+        Equipment equipment = equipmentRepository.findByCode(code);
+        if (equipment == null) {
+            throw new CustomException("There is no equipment with code " + code + ".", HttpStatus.NOT_ACCEPTABLE);
+        }
+
+        Asset asset = assetRepository.getAsset(requesterName, code);
+        if (asset == null) {
+            throw new CustomException("The user has no asset named " + code + " in application.",
+                    HttpStatus.NOT_ACCEPTABLE);
+        }
+
+        if (asset.getAmount() < amount) {
```

Comment:
Negative value is acceptable here. May be we have to check for this and throw exception when necessary

Reply:
New checker added.